pkg/dns/dns.go

```diff
@@ -0,0 +1,417 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dns
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/golang/glog"
+	"hash/fnv"
+	"net"
+	"strings"
+	"time"
+
+	etcd "github.com/coreos/etcd/client"
+	skymsg "github.com/skynetservices/skydns/msg"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/endpoints"
+	kcache "k8s.io/kubernetes/pkg/client/cache"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	kframework "k8s.io/kubernetes/pkg/controller/framework"
+	kselector "k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/util/validation"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+const (
+	kubernetesSvcName = "kubernetes"
+
+	// A subdomain added to the user specified domain for all services.
+	serviceSubdomain = "svc"
+
+	// A subdomain added to the user specified dmoain for all pods.
+	podSubdomain = "pod"
+
+	// Resync period for the kube controller loop.
+	resyncPeriod = 30 * time.Minute
+)
+
+type KubeDNS struct {
+	kubeClient *kclient.Client
+	// DNS domain name.
+	domain string
+	// A cache that contains all the endpoints in the system.
+	endpointsStore kcache.Store
+	// A cache that contains all the services in the system.
+	servicesStore     kcache.Store
+	cache             *TreeCache
+	domainPath        []string
+	eController       *kframework.Controller
+	serviceController *kframework.Controller
+}
+
+func NewKubeDNS(client *kclient.Client, domain string) *KubeDNS {
+	kd := &KubeDNS{
+		kubeClient: client,
+		domain:     domain,
+		cache:      NewTreeCache(),
+		domainPath: reverseArray(strings.Split(strings.TrimRight(domain, "."), ".")),
+	}
+	kd.setEndpointsStore()
+	kd.setServicesStore()
+	return kd
+}
+
+func (kd *KubeDNS) Start() {
+	go kd.eController.Run(wait.NeverStop)
+	go kd.serviceController.Run(wait.NeverStop)
+	// Wait synchronously for the Kubernetes service and add a DNS record for it.
+	// TODO (abshah) UNCOMMENT AFTER TEST COMPLETE
+	//kd.waitForKubernetesService()
```

Comment:
It's because the liveness/readiness probe looks up the "kubernetes" service, and it wasn't passing in < 5m causing kube-up timeouts in the e2e suite (https://github.com/kubernetes/kubernetes/issues/21940). so waiting for it synchronously gives us a binary signal if:
1. Kubernetes service was not created by master, not transmitted down the watch etc 
2. DNS is flaking in spite of having the kubernetes service